tools/server/utils.hpp

```diff
@@ -935,3 +982,277 @@ static std::vector<common_adapter_lora_info> parse_lora_request(
 
     return lora;
 }
+
+//
+// utils for interacting with libmtmd
+// (may need to refactor in near future)
+//
+
+/**
+ * server_tokens is a helper to manage the input tokens and image for the server.
+ * it is made this way to simplify the logic of KV cache management.
+ */
+struct server_tokens {
+    bool has_mtmd = false;
+
+private: // disallow accessing these members directly, risking out-of-sync
+
+    // map a **start** position in tokens to the image chunk
+    std::unordered_map<llama_pos, mtmd_input_chunk> map_pos_to_image;
+
+    // list of tokens
+    // it can include LLAMA_TOKEN_NULL, which is used to indicate a token that is not a text token
+    // a mtmd_input_chunk can occupy multiple tokens, one llama_token per **position**
+    // important: for models using mrope, an image can contain multiple tokens but will use only one **position**
+    llama_tokens tokens;
+
+    // for ex. with input of 5 text tokens and 2 images:
+    //      [0] [1] [2] [3] [4] [img0] [img0] [img0] [img1] [img1]
+    // pos  0   1   2   3   4   5      6      7      8      9
+    // map_pos_to_image will contain: {5, img0}, {8, img1}
```

Comment:
@ggerganov This is the new way that I propose. To address the risk of desync between `tokens` and `map_pos_to_image`, these members are private.

To provide access to the list of underlaying `tokens`, I added a function `get_text_tokens()`. It will still be a copy in case the caller wants to manipulate the list, an example can be found in the logic of context shift and prompt truncation (tbh I think we should re move the prompt truncation altogether). However, I think these copies can easily be removed by adding functions like `server_tokens::remove_first(n)`